### PR TITLE
 Icon for debugger in the toolbar

### LINKED ISSUE #16594: Icon for debugger in the toolbar
It would be great to have a different icon for the debugger in the toolbar. 
In Pharo12, we always have the same icons

image

In the above example, the same icon for a playground and a debugger.
It forces to read to retrieve its debugger

### PR DESCRIPTION
**Pull request by badetitou**
In the past, we were use to have the icon of the window in the toolbar.

## Issue summary 

For instance, in Pharo 10, we had the icon for the debugger (see screenshot)
image

But, in Pharo12. we lose this icon (see second screenshot)
image

This PR aims to add it back as presented in the last screenshot
image

## Fix detail

To fix the issue. I first discovered that `TaskbarItemMorph` is created from a `TaskbarTask`, but the model of `TaskbarItemMorph` is the actual window morph and not the `TaskbarTask` object. However, the window might not have all the correct information such as the correct icon. However, the `TaskbarTask` has a reference to the window morph, the label of the `TaskbarItemMorph`, the correct label, and the state of the window morph.
So, I concluded that `TaskbarTask` is the perfect object for being the model of the `TaskbarItemMorph`. I thus modify the code to make `TaskbarTask` the model of `TaskbarItemMorph`.
Then, I fixed the methods of `TaskbarTask` and `TaskbarItemMorph` in a way that they call the correct method and tada.

On my computer, I have the correct icon, I have pre-visualisation of windows, and I can open/close windows. I also have the correct label color with minimized and maximized windows in the taskbar.

I do not know where to had tests for this kind of feature, but I can have a look if someone has idea (I dunno if it was  tested) 

fix #16594

### CODE REVIEW

**Comment by Rinzwind** [comment]
The icons seem to have gone missing due to the changes in [commit b25824e010604f57](https://github.com/pharo-project/pharo/commit/b25824e010604f57209386a471cc96f5c22c5f6e) and [Spec commit 1e0ba07e672b77af](https://github.com/pharo-spec/Spec/commit/1e0ba07e672b77af47f1c416fd0a83f950b35cf0): [`#initializeFor:` on TaskbarItemMorph](https://github.com/pharo-project/pharo/commit/b25824e010604f57209386a471cc96f5c22c5f6e#diff-601a7bdaecbe259e52f4ac15cd2776f3fd221c75dc65336f9b67f18f3a996a76) was changed to send `#taskbarIconName` instead of `#taskbarIcon`, and the implementation of [`#taskbarIcon` on SystemWindow](https://github.com/pharo-project/pharo/commit/b25824e010604f57209386a471cc96f5c22c5f6e#diff-bb8d466ada99403f7f1c4ad63d47a3bddaf9732f63a53068b33ad5981327b76d) was replaced by an implementation of `#taskbarIconName` that sends the same message, but the implementation of [`#taskbarIcon` on SpWindowPresenter](https://github.com/pharo-spec/Spec/commit/1e0ba07e672b77af47f1c416fd0a83f950b35cf0#diff-3a2f07f363906e62a1052f79d29ef2de889d66a4b74b69b3e226450175e79503) was not replaced by an implementation of `#taskbarIconName`. Implementing `#taskbarIconName` on SpWindowPresenter as follows restores the icon for a debugger (it’s only a partial fix as it ignores the send of `#windowIcon` in `#taskbarIcon` for which there is no equivalent `#windowIconName` yet):

```smalltalk
taskbarIconName

	^ self presenter
		ifNil: [ super taskbarIconName ]
		ifNotNil: #taskbarIconName
```

Note that the change to [`#initializeFor:` on TaskbarItemMorph](https://github.com/pharo-project/pharo/pull/16661/commits/09f40b6ecb361d2f9d6db3a4bc4e9331bfd24713#diff-601a7bdaecbe259e52f4ac15cd2776f3fd221c75dc65336f9b67f18f3a996a76L68-R68) in this pull request reverses the use of the FormSet for the icon introduced in pull request #16316.

**Comment by badetitou** [comment]
Thanks for the explanation.

Do you think `TaskbarTask` should not be the model of the `TaskbarItemMorph`?
Maybe we could modify `TaskbarTask` by changing the `icon` attribute by `iconName`, and so keeping the usage of FormSet?

**Comment by Rinzwind** [comment]
I would suggest to maybe first fix the missing `#taskbarIconName` on SpWindowPresenter and to then check whether it would still make sense to change the model passed to the TaskbarItemMorph.

A change in commit 09f40b6ecb361d2f that is not so clear to me is the removal of the send of `#taskbarButtonFor:` from `#taskbarButtonFor:` on TaskbarTask which seems to not take into account that there are other implementations of `#taskbarButtonFor:` besides the one on SystemWindow. The implementation of `#performAction:` on TaskbarItemMorph seems to be a copy of the one inherited from PluggableButtonMorph with `model` replaced by `model morph`, that may need to be abstracted better.

**Comment by badetitou** [comment]
I try to do this in an image based on the code of `#taskbarIcon`. But it does not fix the issue. 

> I would suggest to maybe first fix the missing #taskbarIconName on SpWindowPresenter and to then check whether it would still make sense to change the model passed to the TaskbarItemMorph.


I agree that `model morph` is strange and we should find a better solution

File: src/Morphic-Base/Object.extension.st [hunk 1]
```diff
@@ -114,7 +114,7 @@ Object >> taskbarIconName [
 	"Answer the icon for the receiver in a task bar
 	or nil for the default."
 
-	^self class taskbarIconName
```
Comment (badetitou) [inline src/Morphic-Base/Object.extension.st:117]: This is only default formatting

**Comment by badetitou** [comment]
I updated the code to use only iconName instead of icon and so to use a iconSetForm.

**Comment by Ducasse** [comment]
So I imagine that now I can merge?

**Comment by Rinzwind** [comment]
@badetitou: Regarding [your third comment](https://github.com/pharo-project/pharo/pull/16661#issuecomment-2149223408), I’m not quite sure what you tried and why that wouldn’t have worked. Try implementing the following on SpWindowPresenter in Pharo 13 build 97:

```smalltalk
windowIconName

	^ self windowIcon ifNotNil: [ :form |
		Smalltalk ui icons icons keyAtValue: form ifAbsent: [ #notFound ] ]

taskbarIconName

	^ self windowIconName ifNil: [
		self presenter
			ifNil: [ super taskbarIconName ]
			ifNotNil: #taskbarIconName ]
```

This implementation of `#windowIconName` is a hack (not to be committed!). I’m assuming that, following the changes in [commit b25824e010](https://github.com/pharo-project/pharo/commit/b25824e010604f57209386a471cc96f5c22c5f6e) and [Spec commit 1e0ba07e67](https://github.com/pharo-spec/Spec/commit/1e0ba07e672b77af47f1c416fd0a83f950b35cf0), the idea is that `#windowIcon:` should be reworked to `#windowIconName:` so that `#windowIconName` would just be an accessor.

One case in which this won’t work is the icon created in `#iconForWindow` on IceTipGitHubRepositoryPanel. That could be solved by adding it to the ThemeIcons as is done in `#iconName` for StRewriterMatchToolPresenter. But that seems like a hack as well, it would be better to have those two icons added to the [icon pack](https://github.com/pharo-project/pharo-icon-packs/tree/svgPack).

**Comment by Rinzwind** [comment]
I opened a pull request to add the two icons to the icon pack: [pharo-icon-packs pull request #13](https://github.com/pharo-project/pharo-icon-packs/pull/13).

**Pull request merged**